File: 3203-find-minimum-diameter-after-merging-two-trees/3203-find-minimum-diameter-after-merging-two-trees.js
var bfs = function (src, adj) {
    const n = adj.length;
    const q = [src];
    const vis = Array(n).fill(false);
    let lst = 0, length = 0;

    while (q.length) {
        const size = q.length;
        for (let i = 0; i < size; i++) {
            const node = q.shift();
            vis[node] = true;
            lst = node;
            for (const child of adj[node]) {
                if (!vis[child]) q.push(child);
            }
        }
        length++;
    }
    return [lst, length];
};

var getDiameter2 = function (edges) {
    const n = edges.length + 1;
    const adj = Array.from({ length: n }, () => []);
    for (const [u, v] of edges) {
        adj[u].push(v);
        adj[v].push(u);
    }
    const [lst] = bfs(0, adj);
    const [, length] = bfs(lst, adj);
    return length - 1;
};

var minimumDiameterAfterMerge = function (edges1, edges2) {
    const d1 = getDiameter2(edges1);
    const d2 = getDiameter2(edges2);
    const d1d2 = Math.ceil(d1 / 2) + Math.ceil(d2 / 2) + 1;
    return Math.max(d1, d2, d1d2);
};

File: 3203-find-minimum-diameter-after-merging-two-trees/3203-find-minimum-diameter-after-merging-two-trees.py
import math
from collections import deque

class Solution:
    def dfs(self, node, parent, adj, dep):
        if dep > self.d:
            self.d = dep
            self.lst = node
        for child in adj[node]:
            if child == parent:
                continue
            self.dfs(child, node, adj, dep + 1)
    
    def getDiameter(self, edges):
        n = len(edges) + 1
        adj = [[] for _ in range(n)]
        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)
        self.d, self.lst = 0, 0
        self.dfs(0, -1, adj, 0)
        self.d = 0
        self.dfs(self.lst, -1, adj, 0)
        return self.d

    def bfs(self, src, adj):
        n = len(adj)
        q = deque([src])
        vis = [False] * n
        lst, length = 0, 0
        while q:
            size = len(q)
            for _ in range(size):
                node = q.popleft()
                vis[node] = True
                lst = node
                for child in adj[node]:
                    if not vis[child]:
                        q.append(child)
            length += 1
        return lst, length

    def getDiameter2(self, edges):
        n = len(edges) + 1
        adj = [[] for _ in range(n)]
        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)
        lst, _ = self.bfs(0, adj)
        _, length = self.bfs(lst, adj)
        return length - 1

    def minimumDiameterAfterMerge(self, edges1, edges2):
        d1 = self.getDiameter2(edges1)
        d2 = self.getDiameter2(edges2)
        d1d2 = math.ceil(d1 / 2.0) + math.ceil(d2 / 2.0) + 1
        return max(d1, d2, d1d2)
